opm/autodiff/StandardWell_impl.hpp

```diff
@@ -2229,4 +2232,88 @@ namespace Opm
             }
         }
     }
+
+
+
+
+    template<typename TypeTag>
+    double
+    StandardWell<TypeTag>::
+    relaxationFactorFraction(const double old_value,
+                             const double dx) {
+        // may adding a tolerance?
+        assert(old_value >= 0. && old_value <= 1.0);
+
+        // to avoid to get negative value or value above 1
+        double relaxation_factor = 1.;
+
+        bool relaxation_avoid_negative = false;
+        bool relaxation_avoid_over_one = false;
+        {
+            const double possible_updated_value = old_value - dx;
+            relaxation_avoid_negative = possible_updated_value < 0.;
+            relaxation_avoid_over_one = possible_updated_value > 1.0;
+
+            if (!relaxation_avoid_negative && !relaxation_avoid_over_one) {
+                return relaxation_factor; // 1.0
+            }
+        }
+
+        // we will not be this step if dx == 0.
+        assert(dx != 0.0);
+
+        if (relaxation_avoid_negative) {
+            relaxation_factor = std::abs(old_value / dx) * 0.95;
+        } else if (relaxation_avoid_over_one) {
+            relaxation_factor = std::abs((1. - old_value) / dx) * 0.95;
+        }
+
+        assert(relaxation_factor >= 0. && relaxation_factor <= 1.);
+        return relaxation_factor;
+    }
+
+
+
+
+    template<typename TypeTag>
+    double
+    StandardWell<TypeTag>::
+    determineRelaxationFractor(const std::vector<double>& primary_variables,
+                               const BVectorWell& dwells)
+    {
+        // TODO: not considering solvent yet
+
+        double relaxation_factor = 1.0;
+
+        if (FluidSystem::phaseIsActive(FluidSystem::waterPhaseIdx)) {
+            const double relaxation_factor_w = relaxationFactorFraction(primary_variables[WFrac], dwells[0][WFrac]);
+            if (relaxation_factor_w < relaxation_factor) {
```

Comment:
Using `relaxation_factor = std::min(relaxation_factor, relaxation_factor_w)` would be a little simpler here.